File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int  a = s1.size();
        int b = s2.size();
        unordered_map<char,int>mp1,mp2;
        for(char ch : s1){
            mp1[ch]++;
        }
        if(s1.size() > s2.size()) return false;
        for(int i = 0; i <a; i++){
            mp2[s2[i]]++;
        }

        if(mp1 == mp2) return true;
        for(int i = a; i < b; i++){
            mp2[s2[i]]++;

            mp2[s2[i - a]]--;

            if(mp2[s2[i - a]] == 0){
                mp2.erase(s2[i - a]);
            }
            if(mp1 == mp2) return true;
        }
        return false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {

        int k = s1.length();

        for (int i = 0; i <= s2.length() - k; i++) {

            HashMap<Character, Integer> mp = new HashMap<>();

            for (char ch : s1.toCharArray()) {
                mp.put(ch, mp.getOrDefault(ch, 0) + 1);
            }

            for (int j = i; j < i + k; j++) {
                char ch = s2.charAt(j);

                if (!mp.containsKey(ch)) {
                    break;
                }

                mp.put(ch, mp.get(ch) - 1);

                if (mp.get(ch) == 0) {
                    mp.remove(ch);
                }
            }

            if (mp.isEmpty()) {
                return true;
            }
        }

        return false;
    }
}